src/couch_views/src/couch_views_reader.erl

```diff
@@ -0,0 +1,185 @@
+                        Opts,
+                        Fun,
+                        KeyAcc1
+                    )
+            end, Acc0, expand_keys_args(Args)),
+
+            #{
+                acc := UserAcc2
+            } = Acc1,
+            {ok, maybe_stop(UserCallback(complete, UserAcc2))}
+        end)
+    catch throw:{done, Out} ->
+        {ok, Out}
+    end.
+
+
+handle_row(_DocId, _Key, _Value, #{skip := Skip} = Acc) when Skip > 0 ->
+    Acc#{skip := Skip - 1};
+
+handle_row(DocId, Key, Value, Acc) ->
+    #{
+        db := TxDb,
+        mrargs := Args,
+        callback := UserCallback,
+        acc := UserAcc0
+    } = Acc,
+
+    BaseRow = [
+        {id, DocId},
+        {key, Key},
+        {value, Value}
+    ],
+
+    Row = BaseRow ++ if not Args#mrargs.include_docs -> []; true ->
+        DocOpts0 = Args#mrargs.doc_options,
+        DocOpts1 = DocOpts0 ++ case Args#mrargs.conflicts of
+            true -> [conflicts];
+            _ -> []
+        end,
+        DocObj = case fabric2_db:open_doc(TxDb, DocId, DocOpts1) of
+            {ok, Doc} -> couch_doc:to_json_obj(Doc, DocOpts1);
+            {not_found, _} -> null
+        end,
+        [{doc, DocObj}]
+    end,
+
+    UserAcc1 = maybe_stop(UserCallback({row, Row}, UserAcc0)),
+    Acc#{acc := UserAcc1}.
+
+
+get_view_id(Lang, Args, ViewName, Views) ->
+    case couch_mrview_util:extract_view(Lang, Args, ViewName, Views) of
+        {map, View, _Args} -> View#mrview.id_num;
+        {red, {_Idx, _Lang, View}} -> View#mrview.id_num
+    end.
+
+
+expand_keys_args(#mrargs{keys = undefined} = Args) ->
+    [Args];
+
+expand_keys_args(#mrargs{keys = Keys} = Args) ->
+    lists:map(fun(Key) ->
+        Args#mrargs{
+            start_key = Key,
+            end_key = Key
+        }
+    end, Keys).
+
+
+mrargs_to_fdb_options(Args) ->
+    #mrargs{
+        start_key = StartKey0,
+        start_key_docid = StartKeyDocId,
+        end_key = EndKey0,
+        end_key_docid = EndKeyDocId,
+        direction = Direction,
+        limit = Limit,
+        skip = Skip,
+        inclusive_end = InclusiveEnd
+    } = Args,
+
+    StartKey1 = if StartKey0 == undefined -> undefined; true ->
+        couch_views_encoding:encode(StartKey0, key)
+    end,
+
+    StartKeyOpts = case {StartKey1, StartKeyDocId} of
+        {undefined, _} ->
+            [];
+        {StartKey1, StartKeyDocId} ->
+            [{start_key, {StartKey1, StartKeyDocId}}]
+    end,
+
+    EndKey1 = if EndKey0 == undefined -> undefined; true ->
+        couch_views_encoding:encode(EndKey0, key)
+    end,
+
+    EndKeyOpts = case {EndKey1, EndKeyDocId, Direction} of
+        {undefined, _, _} ->
+            [];
+        {EndKey1, <<>>, rev} when not InclusiveEnd ->
+            % When we iterate in reverse with
+            % inclusive_end=false we have to set the
+            % EndKeyDocId to <<255>> so that we don't
+            % include matching rows.
+            [{end_key_gt, {EndKey1, <<255>>}}];
```

Comment:
I am curious how hard it is to switch to fdb primitives here. We have [`erlfdb_key:first_greater_than/1`](https://github.com/cloudant-labs/couchdb-erlfdb/tree/master/src#L46).

Reply:
I think we would need another function in `erlfdb_key` to support this

Reply:
@iilyak You'll want to check out the code in `fabric_fdb:get_fold_opts/2` which translates between CouchDB semantics and FoundationDB semantics.